### PR TITLE
Introduce defaults in the auth section

### PR DESCRIPTION
**Pull request by chrzaszcz**
**Goal:** POC of defaults for nested options in an optional section to see how they would look like and decide if we want to proceed with this approach for the whole configuration file.

**Scope:** Defaults for `methods`, `sasl_mechanisms` and `sasl_external` in the `auth` section.

**Motivation** - advantages of defaults:
- Easy to find the default value in the code.
- Simpler code for fetching the values.
- No need to recompute larger values like the default list of SASL mechanisms every time they are fetched.

**Challenges:**
- Optional sections (like `auth`) need to be always included to make the defaults present. This cannot be a universal rule as most sections need to be included only when particular features are enabled.
- Small tests: nested values need to be checked selectively instead of comparing the whole configuration terms. However, this seems like a general improvement and the test cases look cleaner afterwards.
- Small tests: defaults need to be inserted into the expected options for the file tests. To fix this, the last commit changes the static `*.options` files into dynamic lists created by Erlang code. This seems to provide more flexibility, but you can check the last commit and see if you like the new, less verbose version more.

See the commit messages for more details.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/93433) / small_tests / 9c68bd827283d455a8cf8b326cc86aef0cfddb18
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3439/93433/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3439/93433/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2021-12-08_18.31.29/lib.mongooseim.logs/run.2021-12-08_18.31.29/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/93462) / small_tests / 0f8e8efa0c59c417f5160777ac075c842cc352b1
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3439/93462/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3439/93462/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2021-12-08_18.37.06/lib.mongooseim.logs/run.2021-12-08_18.37.06/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/93517) / small_tests / 932b5287c2c13ec01d9b20c6388848511b693ff1
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3439/93517/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3439/93517/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2021-12-09_08.00.03/lib.mongooseim.logs/run.2021-12-09_08.00.04/suite.log.html)

**Review by Premwoik — APPROVED** [review]
Looks good to me :)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/93708) / small_tests / 4d10fc79989d7012cca714cf00585516b08c2043
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3439/93708/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3439/93708/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2021-12-09_15.36.30/lib.mongooseim.logs/run.2021-12-09_15.36.30/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/93735) / small_tests / 2a98f47da904bec09953fa6581a824b26968c073
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3439/93735/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3439/93735/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2021-12-09_15.40.47/lib.mongooseim.logs/run.2021-12-09_15.40.48/suite.log.html)

**Review by NelsonVides — COMMENTED** [review]
Hmm... don't know if it is too late for this idea, but the fact that we'd have defaults for entire sections seems to make some things more verbose 🤔 Didn't we consider setting these defaults per-option, like, in https://github.com/esl/MongooseIM/blob/2a98f47da904bec09953fa6581a824b26968c073/include/mongoose_config_spec.hrl#L22, it could have a new element like
```erlang
-record(option, {type :: mongoose_config_spec:option_type(),
                 default :: some_type(),
                 validate = any :: mongoose_config_validator:validator(),
                 process :: undefined | mongoose_config_parser_toml:processor(),
                 wrap = default :: mongoose_config_spec:wrapper()}).
```
Probably there's some reasons not to do this one, like, Erlang would just create all the records with `undefined` there, but that's probably possible to work around. Maybe the work-around would be much more tedious and we should discard that idea?
Asking for some thoughts 🙂

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/93794) / small_tests / 953b3880df6498198bed0820fe1f35e4329994d0
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3439/93794/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3439/93794/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2021-12-09_17.58.56/lib.mongooseim.logs/run.2021-12-09_17.58.57/suite.log.html)

**Comment by chrzaszcz** [comment]
> Hmm... don't know if it is too late for this idea, but the fact that we'd have defaults for entire sections seems to make some things more verbose 🤔 Didn't we consider setting these defaults per-option

Yes I considered it and implemented one whole section with it (general) and it was worse because the `default` would work if we are in a section, but would make no sense if we are not in a section, so you would always need to go one step up (which is tricky) to verify if it makes sense or not, I quickly got lost there. It was also more difficult to code, the resulting code was more complex. What is more the `required` keys are also on the section level, so it makes sense to have `defaults` at the same level.

File: src/sasl/cyrsasl_external.erl [hunk 1]
```diff
@@ -105,13 +105,7 @@ authorize(Creds) ->
...
 get_verification_list(Creds) ->
     HostType = mongoose_credentials:host_type(Creds),
-    case mongoose_config:get_opt([{auth, HostType}, sasl_external], [standard]) of
-        [] -> [standard];
-        List when is_list(List) -> List;
-        standard -> [standard];
-        use_common_name -> [standard, common_name];
-        allow_just_user_identity -> [standard, auth_id]
-    end.
+    mongoose_config:get_opt([{auth, HostType}, sasl_external]).
```
Comment (NelsonVides) [inline src/sasl/cyrsasl_external.erl:108]: Yeah these kinds of things improve so much 😄

**Review by NelsonVides — APPROVED** [review]
All right, the design makes sense, thanks for this work 👌🏽

**Pull request merged**